File: user-service/src/main/java/com/phorus/userservice/services/impls/UserServiceImpl.java
package com.phorus.userservice.services.impls;

import com.phorus.userservice.exceptions.ResourceNotFoundException;
import com.phorus.userservice.model.dbentities.User;
import com.phorus.userservice.model.dtos.UserDTO;
import com.phorus.userservice.repositories.UserRepository;
import com.phorus.userservice.services.PetService;
import com.phorus.userservice.services.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class UserServiceImpl implements UserService {

    private final UserRepository userRepository;
    private final PetService petService;

    @Autowired
    public UserServiceImpl(UserRepository userRepository, PetService petService) {
        this.userRepository = userRepository;
        this.petService = petService;
    }

    @Override
    public User findById(Long userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + userId));
    }

    @Override
    public Long create(User user) {
        return userRepository.save(user).getId();
    }

    @Override
    public void update(Long userId, UserDTO userDTO) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + userId));

        if (userDTO.getName() != null && !userDTO.getName().isEmpty()) {
            user.setName(userDTO.getName());
        }

        if (userDTO.getAge() != null) {
            user.setAge(userDTO.getAge());
        }

        userRepository.save(user);
    }

    @Override
    public void deleteById(Long userId) {
        userRepository.deleteById(userId);
    }


    @Override
    public Long getAmountOfPets(Long userId) {
        return petService.getAmountOfPets(userId);
    }
}

File: user-service/src/main/kotlin/com/phorus/userservice/services/impls/UserServiceImpl.kt
package com.phorus.userservice.services.impls;

import com.phorus.userservice.exceptions.ResourceNotFoundException
import com.phorus.userservice.model.dbentities.User
import com.phorus.userservice.model.dtos.UserDTO
import com.phorus.userservice.repositories.UserRepository
import com.phorus.userservice.services.PetService
import com.phorus.userservice.services.UserService
import org.springframework.stereotype.Service

@Service
class UserServiceImpl(
    private val userRepository: UserRepository,
    private val petService: PetService,
) : UserService {

    override suspend fun findById(userId: Long): User =
        userRepository.findById(userId)
            ?: throw ResourceNotFoundException("User not found with id: $userId")

    override suspend fun create(user: User): Long =
        userRepository.save(user).id as Long

    override suspend fun update(userId: Long, userDTO: UserDTO) {
        val user = userRepository.findById(userId)
            ?: throw ResourceNotFoundException("User not found with id: $userId")

        user.apply {
            userDTO.name?.let { this.name = it }
            userDTO.age?.let { this.age = it }
        }

        userRepository.save(user)
    }

    override suspend fun deleteById(userId: Long) =
        userRepository.deleteById(userId)

    override suspend fun getAmountOfPets(userId: Long): Long =
        petService.getAmountOfPets(userId)
}
